File: 2016/solutions/Day15.java
public class Day15 {
    public static void main(String[] args) {
        int time = 0;
        boolean part_one = false;

        while (true) {
            if ((16 + time) % 17 == 0 &&
                (1 + time) % 3 == 0 &&
                (7 + time) % 19 == 0 &&
                (6 + time) % 13 == 0 &&
                time % 7 == 0 &&
                (1 + time) % 5 == 0) {
                if (!part_one) {
                    //part one
                    System.out.println(time);
                    part_one = true;
                }

                if ((7 + time) % 11 == 0) {
                    //part two
                    System.out.println(time);
                    break;
                }
            }
            time++;
        }
    }
}

File: 2017/solutions/day15.js
function* a_gen() {
    let val = 116;
    while (true) {
        val = (val * 16807) % 2147483647;
        yield val;
    }
}

function* b_gen() {
    let val = 299;
    while (true) {
        val = (val * 48271) % 2147483647;
        yield val;
    }
}

const a = a_gen();
const b = b_gen();
let a_vals = [], b_vals = [], count = 0;

let test_val = (val, is_a=true) => {
    if (is_a) {
        if (val % 4 === 0) a_vals.push(val);
    } else if (val % 8 === 0) b_vals.push(val);
    return val;
}

for (let i = 0;  i < 40000000; i++) {
    if (test_val(a.next().value) % 2**16 === test_val(b.next().value, false) % 2**16) count++;
}

//part one
console.log(count);

while (a_vals.length < 5000000) test_val(a.next().value);
while (b_vals.length < 5000000) test_val(b.next().value, false);

//part two
console.log(a_vals.filter((x, i) => x % 2**16 === b_vals[i] % 2**16).length);